### PR TITLE
SPLAT-1582: Multi vCenter Support

### PR DESCRIPTION
**Pull request by vr4manta**
[SPLAT-1582](https://issues.redhat.com//browse/SPLAT-1582)

### Changes
- Creating support for multiple vCenters in a vSphere installation

### CODE REVIEW

File: enhancements/vsphere-multi-vcenter.md [hunk 1]
```diff
@@ -0,0 +1,839 @@
...
+### Goals
+```
+Summarize the specific goals of the proposal. How will we know that
+this has succeeded?  A good goal describes something a user wants from
+their perspective, and does not include the implementation details
+from the proposal.
+```
+
+- During installation, all nodes are created in all defined vCenters.
+  Rational: multiple vCenters is another twist on failure domains.  A new vCenter
```
Comment (rvanderp3) [inline enhancements/vsphere-multi-vcenter.md:66]:
```suggestion
  Rationale: multiple vCenters is another twist on failure domains.  A new vCenter 
```

File: enhancements/vsphere-multi-vcenter.md [hunk 2]
```diff
@@ -0,0 +1,839 @@
...
+```
+
+- During installation, all nodes are created in all defined vCenters.
+  Rational: multiple vCenters is another twist on failure domains.  A new vCenter 
+  will appear as a new failure domain to the OCP cluster.  Control plane and 
+  compute nodes must be able to be assigned to any FD.
+
+- OCP clusters will be enhanced to leverage new yaml cloud provider config for
+  vSphere.
+  Rational: The ini configuration has been deprecated and the newer yaml format
```
Comment (rvanderp3) [inline enhancements/vsphere-multi-vcenter.md:72]:
```suggestion
  Rationale: The ini configuration has been deprecated and the newer yaml format
```

File: enhancements/vsphere-multi-vcenter.md [hunk 3]
```diff
@@ -0,0 +1,839 @@
...
+  supports multiple vCenters.
+
+- When updating infrastructure after initial installation, the cluster should be
+  able to accept the newly defined failure domain which points leverages a new 
+  vCenter.  
+
+- Updating the cloud provider config from ini to yaml will be supported.
+  Rational: Existing clusters wish to take advantage of migrating loads to a new
```
Comment (rvanderp3) [inline enhancements/vsphere-multi-vcenter.md:80]:
```suggestion
  Rationale: Existing clusters wish to take advantage of migrating loads to a new
```

File: enhancements/vsphere-multi-vcenter.md [hunk 4]
```diff
@@ -0,0 +1,839 @@
...
+  for the new vCenter.
+
+### Non-Goals
+
+What is out of scope for this proposal? Listing non-goals helps to
+focus discussion and make progress. Highlight anything that is being
+deferred to a later phase of implementation that may call for its own
+enhancement.
+
+- Updating cloud config to yaml format for existing clusters (upgrading OCP to 4.17+)
```
Comment (rvanderp3) [inline enhancements/vsphere-multi-vcenter.md:92]: Should we call out that only clusters installed on 4.17(or the y-stream where this is GA'ed) or later can use this feature?

File: enhancements/vsphere-multi-vcenter.md [hunk 5]
```diff
@@ -0,0 +1,839 @@
...
+Aware config option for max size.
+
+Example:
+```go
+// +kubebuilder:validation:MinItems=0
+// +openshift:validation:FeatureGateAwareMaxItems:featureGate="",maxItems=1
+// +openshift:validation:FeatureGateAwareMaxItems:featureGate=VSphereMultiVCenters,maxItems=3
+// +listType=atomic
+// +optional
+VCenters []VSpherePlatformVCenterSpec `json:"vcenters,omitempty"`
```
Comment (rvanderp3) [inline enhancements/vsphere-multi-vcenter.md:117]: Should we consider a webhook to block updates if the cluster doesnt have a yaml cloud config?

File: enhancements/vsphere-multi-vcenter.md [hunk 6]
```diff
@@ -0,0 +1,839 @@
...
+configured in the infrastructure resource.  We will be controlling this via
+a new feature (VSphereMultiVCenters) and will have different CRDs installed
+based on this gate.
+
+Initially, the plans are to allow a max of 3 vCenters to be configured when the
+feature gate is enabled.  The way we are going to control this is by adding new
+control annotations to the model objects.
+
+The OpenShift controller tools will be enhanced to allow a new Feature Gate 
+Aware config option for max size.
```
Comment (JoelSpeed) [inline enhancements/vsphere-multi-vcenter.md:108]: This was already done right?

File: enhancements/vsphere-multi-vcenter.md [hunk 7]
```diff
@@ -0,0 +1,839 @@
...
+[Labels]
+topology-categories = openshift-zone,openshift-region
+```
+
+Next the operator was enhanced to be able to support using the upstream vSphere 
+YAML cloud provider config format.  There is some logic that uses our old legacy 
+style config.  To preserve this, we created a wrapper config object that attempts
+to load the cloud provider config as either INI or YAML.  If its INI, we will 
+also store the INI data into the `LegacyConfig` field so we can access it in 
+certain situations.
```
Comment (JoelSpeed) [inline enhancements/vsphere-multi-vcenter.md:636]: We have had people confused by the errors from trying to read old vs new, is there anything we can do to mitigate these errors from being produced?

**Review by JoelSpeed — COMMENTED** [review]
Does spanning multiple vCenters introduce anything new that we need to consider with respect to the security of the cluster?

File: enhancements/vsphere-multi-vcenter.md [hunk 8]
```diff
@@ -0,0 +1,839 @@
...
+This leads us to the new checker being added to help the administrator detect when
+such a config mistake may have occurred.  The new check will validate the configs
+in infrastructure and the cloud config to make sure both are configured for the
+same vCenters.  It also makes sure that all failure domains are referencing valid
+vCenters that should be defined in the vCenter section of the infrastructure cluster
+resource.
+
+### Multiple vCenters Configured as Day 2
+
+NOTE: This section is placeholder for future design / work.
```
Comment (jsafrane) [inline enhancements/vsphere-multi-vcenter.md:698]: What do we actually support In OCP 4.17? It is possible to edit `Infrastructure` after installation and all the operators try to respect it. But is it supported? Should we block such edits until we figure out what we actually support?

File: enhancements/vsphere-multi-vcenter.md (hunk 4 above)
Comment (vr4manta) [reply enhancements/vsphere-multi-vcenter.md:92]: Sounds good.  I'll update that.

File: enhancements/vsphere-multi-vcenter.md (hunk 5 above)
Comment (vr4manta) [reply enhancements/vsphere-multi-vcenter.md:117]: We recently updated the CRD to block changes once this field is set in 4.17.  Once we finish migration testing and fixes, the CRD will be updated to not have those blockers anymore.  Right now I do believe the cluster admin will have to manually migrate this config.  I am hoping we can maybe create an automated process for this.

File: enhancements/vsphere-multi-vcenter.md (hunk 7 above)
Comment (vr4manta) [reply enhancements/vsphere-multi-vcenter.md:636]: I am working with upstream to fix these outputs: https://github.com/kubernetes/cloud-provider-vsphere/pull/1225

**Review by JoelSpeed — COMMENTED** [review]
How well does this reflect reality, and what is the state of the feature? If the feature is shipped, we should aim to get this merged asap

The future work that's outline is very sparse at the moment, should it be commented so that it's not rendered until it's actually been discussed and fleshed out?

**Pull request closed without merging**